File: 2024/Tools/Point.cs
namespace AdventOfCode2024.Tools;

record Point(int X, int Y) 
{    
    public static Point operator+(Point a, Point b) =>
        new(a.X + b.X, a.Y + b.Y);

    public static Point operator-(Point a, Point b) =>
        new(a.X - b.X, a.Y - b.Y);

    public static Point operator*(Point a, int mul) =>
        new(a.X * mul, a.Y * mul);

    public override string ToString() => $"{X},{Y}";
};

File: 2015/tools/point.cpp
#include "point.hpp"

using namespace AoC2015::Tools;

Point::Point() : x(0), y(0) {}

Point::Point(int _x, int _y) : x(_x), y(_y) {}

Point Point::operator+(const Point &p) const
{
    return Point(x + p.x, y + p.y);
}

Point &Point::operator+=(const Point &p)
{
    x += p.x;
    y += p.y;
    return *this;
}

Point Point::operator-(const Point &p) const
{
    return Point{x - p.x, y - p.y};
}

bool Point::operator==(const Point &p) const
{
    return x == p.x && y == p.y;
}

bool Point::operator!=(const Point &p) const
{
    return !(*this == p);
}

size_t std::hash<Point>::operator()(const Point &p) const
{
    return hash<size_t>()(p.x & p.y) ^ hash<size_t>()(p.x | p.y);
}
